app/models/sync/location_sync.rb

```diff
@@ -0,0 +1,163 @@
+# frozen_string_literal: true
+
+module Sync
+  # this class expects a locations payload (according to a predefined structure) and a map id
+  # the goal is to sync the locations from the payload with the ones that exist in the database
+  # 1) check if location with :id exists, if not add, if yes update info if necessary
+  # 2) delete all locations that were not included in this payload
+
+  class LocationSync
+    def initialize(locations_payload, map_id)
+      @locations_payload = locations_payload
+      @map = Map.find(map_id)
+      @locations_db = @map.locations.includes(:opening_times)
+
+      @stale_locations_external_ids = calc_stale_locations
+      @locations_to_insert = []
+      @locations_to_update = []
+      @opening_times_to_update = []
+      @opening_times_to_destroy = []
+    end
+
+    def synchronize!
+      @locations_payload.each { |lp| categorize_location(lp) }
+      update_database!
+    end
+
+    def validate!
+      # validate without updating database so that we can return errors in response if necessary
+      @locations_payload.each { |lp| categorize_location(lp) }
+    end
+
+    def update_database!
```

Comment:
Each of these steps can fail, leading to a potentially inconsistent state. For example, the deletions of locations might work, but not the import. Or the import goes through, but only partially. 

I think it's best to wrap all of these database queries under an ActiveRecord::Transaction https://api.rubyonrails.org/classes/ActiveRecord/Transactions/ClassMethods.html

Reply:
Resolving, it's now obsolete with "wipe & create".